apps/emqx/src/emqx_limiter/src/emqx_limiter_shared/emqx_limiter_shared.erl

```diff
@@ -61,101 +98,264 @@
     {emqx_limiter:name(), emqx_limiter:options()}
 ]) -> ok | {error, term()}.
 create_group(Group, LimiterConfigs) when length(LimiterConfigs) > 0 ->
-    {Names, _} = lists:unzip(LimiterConfigs),
-    ChildSpec = child_spec(Group, Names),
-    start_child(emqx_limiter_shared_sup, ChildSpec).
+    Size = length(LimiterConfigs),
+    MiniTokensCRef = counters:new(Size, [write_concurrency]),
+    LastTimeARef = atomics:new(Size * 2, []),
+    NowUs = now_us_monotonic(),
+    Buckets = make_buckets(LimiterConfigs, MiniTokensCRef, LastTimeARef, NowUs),
+    ok = emqx_limiter_bucket_registry:insert_buckets(Group, Buckets).
 
 -spec delete_group(emqx_limiter:group()) -> ok | {error, term()}.
 delete_group(Group) ->
-    case supervisor:terminate_child(emqx_limiter_shared_sup, Group) of
-        ok ->
-            supervisor:delete_child(emqx_limiter_shared_sup, Group);
-        {error, not_found} ->
-            ok
-    end.
+    ok = emqx_limiter_bucket_registry:delete_buckets(Group).
 
 -spec update_group(emqx_limiter:group(), [{emqx_limiter:name(), emqx_limiter:options()}]) ->
-    ok | no_return().
-update_group(Group, _LimiterConfigs) ->
-    ok = emqx_limiter_allocator:update(Group).
-
--spec child_spec(emqx_limiter:group(), [emqx_limiter:name()]) ->
-    supervisor:child_spec().
-child_spec(Group, Names) ->
-    #{
-        id => Group,
-        start => {emqx_limiter_allocator, start_link, [Group, Names]},
-        restart => permanent,
-        shutdown => 5000,
-        type => worker,
-        modules => [?MODULE, emqx_limiter_allocator]
-    }.
+    ok.
+update_group(_Group, _LimiterConfigs) ->
+    ok.
 
 -spec connect(emqx_limiter:id()) -> emqx_limiter_client:t().
 connect({Group, Name}) ->
-    emqx_limiter_client:new(
-        ?MODULE,
-        _State = {Group, Name}
-    ).
+    case emqx_limiter_bucket_registry:find_bucket({Group, Name}) of
+        undefined ->
+            error({bucket_not_found, {Group, Name}});
+        #{last_time_aref := LastBurstTimeARef, last_burst_time_index := LastBurstTimeIndex} =
+                BucketRef ->
+            Options = emqx_limiter_registry:get_limiter_options({Group, Name}),
+            emqx_limiter_client:new(
+                ?MODULE,
+                _State = #{
+                    limiter_id => {Group, Name},
+                    bucket_ref => BucketRef,
+                    last_burst_time => atomics:get(LastBurstTimeARef, LastBurstTimeIndex),
+                    mode => calc_mode(Options),
+                    options => Options
```

Comment:
the options are taken from perrsistent term, why make a copy of it into the state ?

Reply:
To avoid recaculating mode, if settings did not change.